### PR TITLE
Feature/vk channel

### PR DESCRIPTION
**Pull request by allancslima**
I am sending this PR to the demand that emerged in issue [#1125](https://github.com/rapidpro/rapidpro/issues/1125).

VK is a Russian social network similar to Facebook and the equivalent of fb page on VK is the community, so integration occurred via communities.

## Setup

Consider `BASE_CONFIG_PATH` = `community page` > `manage` > `settings` > `API usage`

First, the important fields captured on RP backend form are:

- `Community access token`: used to make almost all requests, created by accessing  `BASE_CONFIG_PATH` > `access tokens` > `create token` (allow access to community messages and photos);
- `Server verification string`: used to verify webhook, Courier needs to return it on server verification event.

After channel creation, webhook URL and verify token must be placed in `BASE_CONFIG_PATH` > `callback API` > `server settings` > `URL` and  `secret key` (first save secret key and then confirm server); the API version must be pointed in version `5.103`.

- That way all requests are validated using the generated verify token;
- Get `server verification string` in this section;
- Still in this section, at `event types` check the event `messages` > `message received`.

## Receive

We need handle two webhook callback types at `/receive`: `server verification` and `new message`. For this, VK provides the Callback API to notify us about various event types, consists of:

- `POST` request to webhook URL with body containing a object in the format: `{"type": "event_type", "object": {...}, "secret": "verify_token"}`
- So, body object to server verification: `{"type": "confirmation", "group_id": community_id, "secret": "verify_token"}`. Needs to return the `server verification string`.
- And body object to received message: `{"type": "message_new", "object": {"message": {...}}, "secret": "verify_token"}`. Needs to return the "ok" string.

See more:

[https://vk.com/dev/callback_api](https://vk.com/dev/callback_api)
[https://vk.com/dev/objects/message](https://vk.com/dev/objects/message)

To message attachments we capture the first media only. The types are _photo_, _sticker_, _graffiti_, _audio_ and _document_. The _video_ type contains some restrictions to be obtained and, unfortunately, we were unable to find the necessary instructions in the docs.

See more:

[https://vk.com/dev/objects/attachments_m](https://vk.com/dev/objects/attachments_m)
[https://vk.com/dev/objects/video](https://vk.com/dev/objects/video)

## Send

Basic usage of VK API:

- `POST` or `GET` request using `https`;
- Request URL format: `https://api.vk.com/method/METHOD_NAME?PARAMETERS&access_token=ACCESS_TOKEN&v=5.103`;
- `ACCESS_TOKEN` in our case is the community access token.

See more:

[https://vk.com/dev/api_requests](https://vk.com/dev/api_requests)
[https://vk.com/dev/methods](https://vk.com/dev/methods)

To send a message we use the `messages.send` method. To attach medias on message we need upload it first to VK media upload server, currently the photos upload only is available via community. The steps to attach a photo to message:

1. Get photo upload server URL: `photos.getMessagesUploadServer` method;
2. Download media: `GET` request to media URL;
3. Upload media: multipart request to the retrieved server URL at step 1;
4. Save uploaded media information: `photos.saveMessagesPhoto` method.

At this point we have a string like "photo123456_789" where "photo" is the media type, "123456" the owner ID (community ID) and "789" the media ID.

Although the photo upload only is available via community the code was written thinking to any media type, if it follows the same pattern, of course.

See more:

[https://vk.com/dev/upload_files](https://vk.com/dev/upload_files)
[https://vk.com/dev/photos.getMessagesUploadServer](https://vk.com/dev/photos.getMessagesUploadServer)
[https://vk.com/dev/photos.saveMessagesPhoto](https://vk.com/dev/photos.saveMessagesPhoto)

### CODE REVIEW

File: handlers/vk/vk.go [hunk 1]
```diff
@@ -0,0 +1,595 @@
...
+	// write required response
+	_, err := fmt.Fprint(w, verificationString)
+
+	return nil, err
+}
+
+// receiveMessage handles new message event
+func (h *handler) receiveMessage(ctx context.Context, channel courier.Channel, w http.ResponseWriter, r *http.Request, payload *moNewMessagePayload) ([]courier.Event, error) {
+	userId := payload.Object.Message.UserId
+	urn, err := urns.NewURNFromParts(urns.VKScheme, strconv.FormatInt(userId, 10), "", "")
```
Comment (allancslima) [inline handlers/vk/vk.go]: PR in `gocommon`: https://github.com/nyaruka/gocommon/pull/26

File: handlers/test.go [hunk 2]
```diff
@@ -194,7 +196,7 @@ func RunChannelSendTestCases(t *testing.T, channel courier.Channel, handler cour
 		t.Run(testCase.Label, func(t *testing.T) {
 			require := require.New(t)
 
-			msg := mb.NewOutgoingMsg(channel, courier.NewMsgID(10), urns.URN(testCase.URN), testCase.Text, testCase.HighPriority, testCase.QuickReplies, testCase.Topic, testCase.ResponseToID, testCase.ResponseToExternalID)
+			msg := mb.NewOutgoingMsg(channel, testCase.MsgID, urns.URN(testCase.URN), testCase.Text, testCase.HighPriority, testCase.QuickReplies, testCase.Topic, testCase.ResponseToID, testCase.ResponseToExternalID)
```
Comment (allancslima) [inline handlers/test.go]: Perhaps this change will break some tests?

File: handlers/test.go [hunk 3]
```diff
@@ -59,6 +59,7 @@ type SendPrepFunc func(*httptest.Server, courier.ChannelHandler, courier.Channel
 type MockedRequest struct {
 	Method       string
 	Path         string
+	RawQuery     string
```
Comment (allancslima) [inline handlers/test.go]: VK requests are made by parameters.

Comment (nicpottier) [reply handlers/test.go]: Do you actually need this though? Isn't the path enough to differentiate between the different calls or am I misunderstanding that?

Comment (allancslima) [reply handlers/test.go]: Yes, I need.

A request to `"/messages.json?access_token=abc123xyz&message=Simple+message""` for example, will be a request with `URL.Path` equal to `"/messages.json"` and `URL.RawQuery` equal to `"access_token=abc123xyz&message=Simple+message"`.

**Pull request merged**